delta/sourcing/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/projections/ProjectionStream.scala

```diff
@@ -138,6 +138,56 @@ object ProjectionStream {
         case v                                    => Task.pure(v)
       }
 
+    def accumulateProgress(initial: ProjectionProgress[A]): Stream[Task, (ProjectionProgress[A], Message[A])] = {
+      stream
+        .mapAccumulate(initial) {
+          case (acc, msg) if msg.offset.gt(initial.offset) => (acc + msg, msg)
+          case (acc, msg)                                  => (acc, msg)
+        }
+    }
+
+    private def persistToProjection(
+        chunk: Chunk[(ProjectionProgress[A], Message[A])],
+        persistProgress: (ProjectionId, ProjectionProgress[A]) => Task[Unit],
+        persistErrors: (ProjectionId, Vector[Message[A]]) => Task[Unit]
+    ): Task[Option[A]] = {
+      val init: (Option[ProjectionProgress[A]], Vector[Message[A]], Option[A]) = (None, Vector.empty, None)
+      val (progress, errors, firstValue)                                       = chunk.foldLeft(init) { case ((_, messages, value), (progress, message)) =>
+        val (error, newValue) = message match {
+          case SuccessMessage(_, _, _, value, warnings) => Option.when(warnings.nonEmpty)(message) -> Some(value)
+          case m                                        => Some(m)                                 -> None
+        }
+        (Some(progress), messages ++ error, value.orElse(newValue))
+      }
+      persistErrors(projectionId, errors) >>
+        progress.fold(Task.unit)(persistProgress(projectionId, _)) >>
+        Task.pure(firstValue)
+    }
+
+    /**
+      * Map over the stream of messages and persist the progress and errors as well as cache progress
+      *
+      * @param initial         where we started
+      * @param persistErrors   how we persist errors
+      * @param persistProgress how we persist progress
+      * @param cacheProgress   how we cache progress
+      * @param config          the config
+      */
+    def persistProgressWithCache(
+        initial: ProjectionProgress[A],
+        persistProgress: (ProjectionId, ProjectionProgress[A]) => Task[Unit],
+        persistErrors: (ProjectionId, Vector[Message[A]]) => Task[Unit],
+        cacheProgress: (ProjectionId, ProjectionProgress[A]) => Task[Unit],
+        config: PersistProgressConfig
+    ): Stream[Task, A] =
+      stream
+        .accumulateProgress(initial)
+        .evalMap { case (progress, message) =>
+          cacheProgress(projectionId, progress).as((progress, message))
```

Comment:
That makes a lot of writes in the cache and Distributed Data is eventually consistent, we really need to persist for every event processed by every view ?

Reply:
this would be a problem indeed; if we need more consistency we would have to switch to the previous way to store the value in memory.

Reply:
If we don't want to persist every message,  we would have to chunk and flatten the stream one more time.
We could also cache the progress in the same place where we record it in the projection , but then the delay would be the same as for the projection.

Reply:
If the progress has already been computed, we can use `debounce(5.seconds)` or something before `.accumulateProgress(initial)`

Reply:
if we use `Ref[Task, A] wouldn't  it we have to run the stream (and indexing) for each view on each node?

Reply:
Yes. That's the penalty to pay for a faster cache.

Reply:
I think `debounce` drops elements, which would mean that some messages are dropped and, in case of errors and warnings, not persisted.

Reply:
otherwise you can do another `groupWithin`, take the latest from the cunks and index it into the cache and then flatten the chunks back into the stream

Reply:
yes, I guess that will be more efficient than writing to cache for every message.